proxy_agent/src/common/http.rs

```diff
@@ -1,478 +1,394 @@
+    while let Some(next) = response.frame().await {
+        let frame = match next {
+            Ok(f) => f,
+            Err(e) => {
+                return Err(std::io::Error::new(
+                    std::io::ErrorKind::Other,
+                    format!("Failed to get next frame from response: {}", e),
+                ))
             }
         };
+        if let Some(chunk) = frame.data_ref() {
+            body_string.push_str(&String::from_utf8_lossy(chunk));
+        }
     }
-
-    Ok(data)
-}
-
-pub fn receive_body(stream: &TcpStream, content_length: usize) -> std::io::Result<Vec<u8>> {
-    let mut reader = BufReader::new(stream);
-    receive_body_internal(&mut reader, content_length)
-}
-
-fn stream_body_internal(
-    mut reader: BufReader<&TcpStream>,
-    mut dest_stream: &TcpStream,
-    len: usize,
-) -> std::io::Result<usize> {
-    let mut received: usize = 0;
-
-    while received < len {
-        match reader.fill_buf() {
-            Ok(d) => {
-                let read = d.len();
-                dest_stream.write_all(d)?;
-                reader.consume(read);
-                received += read;
-            }
-            Err(_e) => {
-                // read timeout, assume no more incoming data in the TcpStream
-                break;
-            }
-        };
+    if json_format {
+        match serde_json::from_str(&body_string) {
+            Ok(t) => Ok(t),
+            Err(e) => Err(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                format!(
+                    "Failed to json deserialize response body from: {} with error {}",
+                    body_string, e
+                ),
+            )),
+        }
+    } else {
+        match serde_xml_rs::from_str(&body_string) {
+            Ok(t) => Ok(t),
+            Err(e) => Err(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                format!(
+                    "Failed to xml deserialize response body from: {} with error {}",
+                    body_string, e
+                ),
+            )),
+        }
     }
-
-    dest_stream.flush()?;
-    Ok(received)
 }
 
-// receive body from source stream and,
-// send to dest stream directly
-pub fn stream_body(
-    source_stream: &TcpStream,
-    dest_stream: &TcpStream,
-    content_length: usize,
-) -> std::io::Result<usize> {
-    let reader = BufReader::new(source_stream);
-    stream_body_internal(reader, dest_stream, content_length)
-}
-
-// forward response from server TcpStream to client TcpStream
-// insert extra headers if have
-pub fn forward_response(
-    server_stream: &TcpStream,
-    mut client_stream: &TcpStream,
-    extra_headers: HashMap<&str, &str>,
-) -> std::io::Result<(Response, usize)> {
-    let mut response_reader = BufReader::new(server_stream);
-
-    let mut response_without_body;
-    match read_response_without_body(&mut response_reader) {
-        Ok(r) => response_without_body = r,
+pub fn get_request(
```

Comment:
Would it make sense to refactor this to `build_request`? At initial glance I thought this was for creating a `GET` request, but after looking at a few examples it was more clear to me that this was for building requests of any http method.